src/plugins/admin/admin.plugin.ts

```diff
@@ -0,0 +1,172 @@
+  'User is not an authorized admin',
+  StatusCodes.UNAUTHORIZED,
+);
+const MissingGithubId = createError(
+  'GAERR002',
+  'Response from Github is missing key `id`',
+  StatusCodes.BAD_REQUEST,
+);
+const MissingGithubUsername = createError(
+  'GAERR003',
+  'Response from Github is missing key `username`',
+  StatusCodes.BAD_REQUEST,
+);
+
+export default async (fastify: FastifyInstance) => {
+  const adminRepository = new AdminRepository();
+
+  fastify.register(fastifySecureSession, {
+    key: Buffer.from(ADMIN_SESSION_SECRET_KEY, 'hex'),
+    cookieName: 'adminSession',
+    cookie: {
+      path: '/admin',
+      secure: PROD,
+      httpOnly: true,
+      maxAge: ADMIN_SESSION_EXPIRATION_IN_SECONDS,
+    },
+    expiry: ADMIN_SESSION_EXPIRATION_IN_SECONDS,
+  });
+  // declare a new authenticator so that passport for admins does not interfere with the user passport strategies
+  // also set the user property to `admin` so that it will be available on `request.admin`
+  const adminPassport = new Authenticator({ userProperty: 'admin' });
+  await fastify.register(adminPassport.initialize());
+  await fastify.register(adminPassport.secureSession());
+
+  adminPassport.use(
+    GITHUB_OAUTH_STRATEGY,
+    new GitHubStrategy(
+      {
+        clientID: GITHUB_CLIENT_ID,
+        clientSecret: GITHUB_CLIENT_SECRET,
+        callbackURL: `${PUBLIC_URL.origin}/admin/auth/github/callback`,
+      },
+      async (accessToken: string, refreshToken: string, profile: GitHubProfile, done) => {
+        const { username, id: githubId } = profile;
+        if (!githubId) {
+          throw new MissingGithubId();
+        }
+        if (!username) {
+          throw new MissingGithubUsername();
+        }
+        // only allow users that are present in the admin table by their username
+        if (await adminRepository.isAdmin(db, githubId)) {
+          console.debug('user is an allowed admin');
+          // update info stored in the table
+          await adminRepository.update(db, { githubId, githubName: username });
+          // You can add admin checks here
+          return done(null, profile);
+        }
+        console.debug('user is not an allowed admin', profile);
+        return done(new NotAnAuthorizedAdmin());
+      },
+    ),
+  );
+
+  adminPassport.registerUserSerializer(async (user: GitHubProfile, req: AdminRequest) => {
+    req.log.info(user);
+    return user.id;
+  });
+  adminPassport.registerUserDeserializer(
+    async (uuid: string, req: AdminRequest): Promise<AdminUser | undefined> => {
+      req.log.info({ uuid }, 'Deserialize admin github id');
+
+      const admin = await adminRepository.get(db, uuid);
+      req.log.info({ admin }, 'Resolved github id to admin user');
+
+      return admin;
+    },
+  );
+
+  fastify.get(
+    '/admin/auth/github',
+    {
+      preValidation: adminPassport.authenticate(GITHUB_OAUTH_STRATEGY, { scope: ['user:email'] }),
+    },
+    async (_request: AdminRequest, _reply) => {},
```

Comment:
Should it do something? Show some success message?

Reply:
From what I understood, the handler here is empty because the strategy does the work. It is invoked when you click the link to the login with github. After that github will call the callback. (There is actually another ping-pong with a code that is entirely handled by the strategy).